controller/system_restore_controller.go

```diff
@@ -0,0 +1,386 @@
+	defer c.logger.Info("Shutting down Longhorn SystemRestore controller")
+
+	if !cache.WaitForNamedCacheSync(c.name, stopCh, c.cacheSyncs...) {
+		return
+	}
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (c *SystemRestoreController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *SystemRestoreController) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.syncSystemRestore(key.(string))
+	c.handleErr(err, key)
+
+	return true
+}
+
+func (c *SystemRestoreController) handleErr(err error, key interface{}) {
+	if err == nil {
+		c.queue.Forget(key)
+		return
+	}
+
+	log := c.logger.WithField("systemRestore", key)
+
+	if c.queue.NumRequeues(key) < maxRetries {
+		log.WithError(err).Warn("Error syncing Longhorn SystemRestore")
+
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	log.WithError(err).Warn("Dropping Longhorn SystemRestore out of the queue", key)
+	c.queue.Forget(key)
+}
+
+func (c *SystemRestoreController) getLoggerForSystemRestore(name string) *logrus.Entry {
+	return c.logger.WithField("systemRestore", name)
+}
+
+func (c *SystemRestoreController) syncSystemRestore(key string) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "%v: fail to sync SystemRestore %v", c.name, key)
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	if namespace != c.namespace {
+		return nil
+	}
+
+	backupTargetClient, err := getBackupTargetClientDefault(c.ds)
+	if err != nil {
+		return err
+	}
+
+	return c.reconcile(name, backupTargetClient)
+}
+
+func (c *SystemRestoreController) reconcile(name string, backupTargetClient engineapi.BackupTargetClientSystemBackupInterface) (err error) {
+	systemRestore, err := c.ds.GetSystemRestore(name)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+
+	log := c.getLoggerForSystemRestore(systemRestore.Name)
+
+	if !c.isResponsibleFor(systemRestore) {
+		return nil
+	}
+
+	if systemRestore.Status.OwnerID != c.controllerID {
+		systemRestore.Status.OwnerID = c.controllerID
+		systemRestore, err = c.ds.UpdateSystemRestoreStatus(systemRestore)
+		if err != nil {
+			// we don't mind others coming first
+			if apierrors.IsConflict(errors.Cause(err)) {
+				return nil
+			}
+			return err
+		}
+		log.Infof("Picked up by SystemRestore Controller %v", c.controllerID)
+	}
+
+	if !systemRestore.DeletionTimestamp.IsZero() {
+		defer func() {
+			if err == nil {
+				err = c.ds.RemoveFinalizerForSystemRestore(systemRestore)
+			}
+		}()
+
+		return c.cleanupSystemRestore(systemRestore)
+	}
+
+	if systemRestore.Status.State == longhorn.SystemRestoreStateCompleted || systemRestore.Status.State == longhorn.SystemRestoreStateError {
+		_, err = c.ds.RemoveSystemRestoreLabel(systemRestore)
+		return err
+	}
```

Comment:
It seems that we will retain the job when the restore is completed or errored...

Reply:
I assume it could be useful to keep the pod available for logs. Guess this could be enhanced later.